### PR TITLE
[Clang] Unify interface for accessing template arguments as written for class/variable template specializations

### PR DESCRIPTION
**Pull request by sdkrystian**
Our current method of storing the template arguments as written for `(Class/Var)Template(Partial)SpecializationDecl` suffers from a number of flaws:
- We use `TypeSourceInfo` to store `TemplateArgumentLocs` for class template/variable template partial/explicit specializations. For variable template specializations, this is a rather unintuitive hack (as we store a non-type specialization as a type). Moreover, we don't ever *need* the type as written -- in almost all cases, we only want the template arguments (e.g. in tooling use-cases).
- The template arguments as written are stored in a number of redundant data members. For example, `(Class/Var)TemplatePartialSpecialization` have their own `ArgsAsWritten` member that stores an `ASTTemplateArgumentListInfo` (the template arguments). `VarTemplateSpecializationDecl` has yet _another_ redundant member "`TemplateArgsInfo`" that also stores an `ASTTemplateArgumentListInfo`. 

This patch eliminates all `(Class/Var)Template(Partial)SpecializationDecl` members which store the template arguments as written, and turns the `ExplicitInfo` member into a `llvm::PointerUnion<const ASTTemplateArgumentListInfo*, ExplicitInstantiationInfo*>` (to avoid unnecessary allocations when the declaration isn't an explicit instantiation). The template arguments as written are now accessed via `getTemplateArgsWritten` in all cases.

This patch is near-complete. I still need to eliminate any instances of the template keyword location being stored for non-explicit-instantiation declarations (since we already store the outer template parameter lists). ~With respect to the clang Index tests, I still need to update the tests such that we check for references to entities in the explicit template argument list of specializations~ (**UPDATE**: now done).

The "most breaking" change is to AST Matchers, insofar that `hasTypeLoc` will no longer match class template specializations (since they no longer store the type as written). I wasn't entirely sure what to replace the tests for this behavior with, so I just commented them out... guidance/feedback would be appreciated on this.

### CODE REVIEW

**Comment by erichkeane** [comment]
>The "most breaking" change is to AST Matchers, insofar that hasTypeLoc will no longer match class template specializations (since they no longer store the type as written). I wasn't entirely sure what to replace the tests for this behavior with, so I just commented them out... guidance/feedback would be appreciated on this.

This part has implications that I'm not sure about, and likely needs to A: be in the 'breaking changes' doc, and B: Have Aaron comment on it.

File: clang/include/clang/AST/DeclTemplate.h [hunk 1]
```diff
@@ -1985,44 +1986,45 @@ class ClassTemplateSpecializationDecl
...
+    if (auto *Info = ExplicitInfo.dyn_cast<ExplicitInstantiationInfo *>())
+      Info->TemplateArgsAsWritten = ArgsWritten;
+    else
+      ExplicitInfo = ArgsWritten;
   }
 
-  /// Gets the location of the template keyword, if present.
-  SourceLocation getTemplateKeywordLoc() const {
-    return ExplicitInfo ? ExplicitInfo->TemplateKeywordLoc : SourceLocation();
+  void setTemplateArgsAsWritten(const TemplateArgumentListInfo &ArgsInfo) {
```
Comment (erichkeane) [inline clang/include/clang/AST/DeclTemplate.h:2017]: Can we 'pair' up the getters/setters?  It makes it easier to read/review if they are 'next' to eachother rather than alphabetized (or whatever order you're going for).

File: clang/include/clang/AST/DeclTemplate.h [hunk 2]
```diff
@@ -2596,23 +2582,24 @@ class VarTemplateSpecializationDecl : public VarDecl,
   llvm::PointerUnion<VarTemplateDecl *, SpecializedPartialSpecialization *>
   SpecializedTemplate;
 
-  /// Further info for explicit template specialization/instantiation.
-  struct ExplicitSpecializationInfo {
-    /// The type-as-written.
-    TypeSourceInfo *TypeAsWritten = nullptr;
+  struct ExplicitInstantiationInfo {
```
Comment (erichkeane) [inline clang/include/clang/AST/DeclTemplate.h:2585]: This is the 2nd struct here that has the same structure.  Should we instead extract this to a 'details' NS and reuse it for all of them?

File: clang/test/AST/ast-dump-template-decls.cpp [hunk 3]
```diff
@@ -222,7 +220,7 @@ int binTempl<int, U>;
 
 template<class U>
 float binTempl<float, U> = 1;
-// CHECK:      VarTemplatePartialSpecializationDecl 0x{{[^ ]*}} <line:{{[0-9]+}}:1, line:{{[0-9]+}}:24> col:7 binTempl 'float' cinit
```
Comment (erichkeane) [inline clang/test/AST/ast-dump-template-decls.cpp:225]: Why are we losing 'const-int' (Is that what cinit means?) here?  Is that perhaps important?

File: clang/test/CXX/drs/dr7xx.cpp [hunk 4]
```diff
@@ -279,7 +279,7 @@ namespace dr727 { // dr727: partial
     // cxx98-11-error@-1 {{variable templates are a C++14 extension}}
     // cxx98-14-error@-2 {{inline variables are a C++17 extension}}
     template<> static inline int v2<T>; // #dr727-v2-T
-    // cxx98-14-error@-1 {{inline variables are a C++17 extension}} 
+    // cxx98-14-error@-1 {{inline variables are a C++17 extension}}
```
Comment (erichkeane) [inline clang/test/CXX/drs/dr7xx.cpp:282]: unrelated change?

File: clang/test/Index/Core/index-source.cpp [hunk 5]
```diff
@@ -285,30 +285,23 @@ template<>
 class SpecializationDecl<int>;
 // CHECK: [[@LINE-1]]:7 | class(Gen,TS)/C++ | SpecializationDecl | c:@S@SpecializationDecl>#I | <no-cgname> | Decl,RelSpecialization | rel: 1
 // CHECK-NEXT: RelSpecialization | SpecializationDecl | c:@ST>1#T@SpecializationDecl
-// CHECK: [[@LINE-3]]:7 | class(Gen,TS)/C++ | SpecializationDecl | c:@S@SpecializationDecl>#I | <no-cgname> | Ref | rel: 0
```
Comment (erichkeane) [inline clang/test/Index/Core/index-source.cpp:288]: Can you explain what is going on here that these lines are being lost?

File: clang/include/clang/AST/DeclTemplate.h (hunk 1 above)
Comment (sdkrystian) [reply clang/include/clang/AST/DeclTemplate.h:2017]: Sure

File: clang/include/clang/AST/DeclTemplate.h (hunk 2 above)
Comment (sdkrystian) [reply clang/include/clang/AST/DeclTemplate.h:2585]: Yeah, I thought about it but I didn't just want to dump it into the clang namespace a la `QualifierInfo`. I'll address this after the test related changes.

File: clang/test/AST/ast-dump-template-decls.cpp (hunk 3 above)
Comment (sdkrystian) [reply clang/test/AST/ast-dump-template-decls.cpp:225]: I removed `cinit` because the output will contain "`explicit_specialization`" before it. 

Despite the output being identical to our current output, this somehow seems to affect whether the tests pass if `--check-prefix=...` is removed from the test command line arguments. If the check prefix arguments are retained and we instead make the expected output lines for `DIRECT`/`SERIALIZED` the same, the tests pass. Merging the two lines into a single `CHECK` line results in the change being necessary for the tests to ass...

File: clang/test/CXX/drs/dr7xx.cpp (hunk 4 above)
Comment (sdkrystian) [reply clang/test/CXX/drs/dr7xx.cpp:282]: Oops, sorry

File: clang/test/Index/Core/index-source.cpp (hunk 5 above)
Comment (sdkrystian) [reply clang/test/Index/Core/index-source.cpp:288]: The "type as written" for a specialization is no longer being indexed as a reference to that specialization. The `TemplateArgumentLoc`s still are, but I need to add the corresponding `CHECK` lines so we know for sure

Comment (sdkrystian) [reply clang/test/Index/Core/index-source.cpp:288]: @erichkeane updated the tests to check for references within `TemplateArgumentLoc`s

File: clang/lib/Serialization/ASTReaderDecl.cpp [hunk 6]
```diff
@@ -2545,10 +2545,12 @@ ASTDeclReader::VisitClassTemplateSpecializationDeclImpl(
   }
 
   // Explicit info.
-  if (TypeSourceInfo *TyInfo = readTypeSourceInfo()) {
-    auto *ExplicitInfo =
-        new (C) ClassTemplateSpecializationDecl::ExplicitSpecializationInfo;
-    ExplicitInfo->TypeAsWritten = TyInfo;
+  if (Record.readBool()) {
+    // FIXME: We don't need to allocate this if ExternLoc and TemplateKeywordLoc
```
Comment (erichkeane) [inline clang/lib/Serialization/ASTReaderDecl.cpp:2549]: This serialization is not expected to be stable.  Could we just change the 'serialization' order here to be `ExternLoc/TemplateKeywordLoc/TemplateArgsList` instead?  That way we could test this 'fixme' that way (in both places!).

Also likely value to extract this whole section into its own function so the two can use the 'same' implementation here.

File: clang/unittests/ASTMatchers/ASTMatchersTraversalTest.cpp [hunk 7]
```diff
@@ -6316,6 +6310,15 @@ TEST(HasAnyTemplateArgumentLoc, BindsToSpecializationWithDoubleArgument) {
...
 
+TEST(HasAnyTemplateArgumentLoc, BindsToExplicitSpecializationWithIntArgument) {
+  EXPECT_TRUE(
+      matches("template<typename T> class A {}; template<> class A<int> {};",
+              classTemplateSpecializationDecl(
+                  hasName("A"), hasAnyTemplateArgument(templateArgument(
+                                    refersToType(asString("int")))))));
+}
+
+#if 0
```
Comment (erichkeane) [inline clang/unittests/ASTMatchers/ASTMatchersTraversalTest.cpp:6321]: Still a WIP, or are these from local testing?

File: clang/unittests/ASTMatchers/ASTMatchersTraversalTest.cpp [hunk 8]
```diff
@@ -6390,6 +6394,7 @@ TEST(HasTemplateArgumentLoc, BindsToSpecializationWithDoubleArgument) {
                               0, hasTypeLoc(loc(asString("double")))))))))));
 }
 
+#if 0
```
Comment (erichkeane) [inline clang/unittests/ASTMatchers/ASTMatchersTraversalTest.cpp:6397]: Same question here :)

File: clang/unittests/ASTMatchers/ASTMatchersTraversalTest.cpp (hunk 7 above)
Comment (sdkrystian) [reply clang/unittests/ASTMatchers/ASTMatchersTraversalTest.cpp:6321]: Oh, hah. I was checking whether we could still match class template specializations based on template arguments. The commented out tests should be replaced with equivalents written using the above syntax (where it makes sense).

So to answer your question, it's from local testing :)

File: clang/include/clang/AST/DeclTemplate.h [hunk 9]
```diff
@@ -1792,23 +1807,11 @@ class ClassTemplateSpecializationDecl
...
-
-    ExplicitSpecializationInfo() = default;
-  };
-
   /// Further info for explicit template specialization/instantiation.
   /// Does not apply to implicit specializations.
-  ExplicitSpecializationInfo *ExplicitInfo = nullptr;
+  llvm::PointerUnion<const ASTTemplateArgumentListInfo *,
+                     ExplicitInstantiationInfo *>
+      ExplicitInfo = nullptr;
```
Comment (ChuanqiXu9) [inline clang/include/clang/AST/DeclTemplate.h:1814]: I spent some time to understand `const ASTTemplateArgumentListInfo *` is only meaningful for partial specializations. Can we reuse ExplicitInstantiationInfo::TemplateArgsAsWrittenTemplateArgsAsWritten? And make `ExplicitInstantiationInfo` to `InstantiationInfo` so that we can use it generally?

**Review by ChuanqiXu9 — COMMENTED** [review]
Sounds good. My  instinct reaction to the title is that you're going to unify the annoying duplicated interfaces for the 5 specialization classes (function specialization, class/var (partial) specializations). But it is still good to merge these things.

File: clang/include/clang/AST/DeclTemplate.h (hunk 9 above)
Comment (sdkrystian) [reply clang/include/clang/AST/DeclTemplate.h:1814]: > I spent some time to understand `const ASTTemplateArgumentListInfo *` is only meaningful for partial specializations. 

Not quite... `const ASTTemplateArgumentListInfo *` is also used for explicit specializations. `ExplicitInstantiationInfo *` (should) only be used for explicit instantiations, I just need to remove any instances of `TemplateKeywordLoc` being set for non-explicit-instantiations specializations (planning to do this prior to merging).

File: clang/include/clang/AST/DeclTemplate.h [hunk 10]
```diff
@@ -2596,27 +2588,14 @@ class VarTemplateSpecializationDecl : public VarDecl,
...
-    /// The location of the template keyword.
-    SourceLocation TemplateKeywordLoc;
-
-    ExplicitSpecializationInfo() = default;
-  };
-
   /// Further info for explicit template specialization/instantiation.
   /// Does not apply to implicit specializations.
-  ExplicitSpecializationInfo *ExplicitInfo = nullptr;
+  llvm::PointerUnion<const ASTTemplateArgumentListInfo *,
```
Comment (erichkeane) [inline clang/include/clang/AST/DeclTemplate.h:2593]: Can we extract this type somewhere universal via a `using` declaration and use it via a useful name?

File: clang/lib/AST/TypePrinter.cpp [hunk 11]
```diff
@@ -1455,21 +1455,17 @@ void TypePrinter::printTag(TagDecl *D, raw_ostream &OS) {
...
-        cast<TemplateSpecializationType>(TAW->getType());
-      Args = TST->template_arguments();
-    } else {
-      const TemplateArgumentList &TemplateArgs = Spec->getTemplateArgs();
-      Args = TemplateArgs.asArray();
-    }
+  if (auto S = dyn_cast<ClassTemplateSpecializationDecl>(D)) {
+    const TemplateParameterList *TParams =
+        S->getSpecializedTemplate()->getTemplateParameters();
+    const auto *TArgAsWritten = S->getTemplateArgsAsWritten();
```
Comment (erichkeane) [inline clang/lib/AST/TypePrinter.cpp:1461]: By coding standard, we can't really use 'auto' here IIRC.

File: clang/lib/AST/TypePrinter.cpp [hunk 12]
```diff
@@ -1455,21 +1455,17 @@ void TypePrinter::printTag(TagDecl *D, raw_ostream &OS) {
...
     IncludeStrongLifetimeRAII Strong(Policy);
-    printTemplateArgumentList(
-        OS, Args, Policy,
-        Spec->getSpecializedTemplate()->getTemplateParameters());
+    if (TArgAsWritten && !Policy.PrintCanonicalTypes)
```
Comment (erichkeane) [inline clang/lib/AST/TypePrinter.cpp:1463]: This exact pattern happens 2x... it would be nice/useful if we could get a helper function to capture this pattern (perhaps an overload of `printTemplateArgumentList`?).

File: clang/lib/Serialization/ASTReaderDecl.cpp (hunk 6 above)
Comment (erichkeane) [reply clang/lib/Serialization/ASTReaderDecl.cpp:2553]: Still waiting on this?

**Review by erichkeane — COMMENTED** [review]
A few more suggestions, also I like Shafik's suggestion on the variable name.

File: clang/lib/AST/TypePrinter.cpp (hunk 12 above)
Comment (erichkeane) [reply clang/lib/AST/TypePrinter.cpp:1463]: Ah, shucks... ok, dont worry about it now then.

File: clang/include/clang/AST/DeclTemplate.h (hunk 10 above)
Comment (erichkeane) [reply clang/include/clang/AST/DeclTemplate.h:2593]: Its to give it a 'useful' descriptive name.  The pointer-union is necessarily opaque as to its meaning, so a name that reflects the name is really helpful when reading the code.

Comment (sdkrystian) [reply clang/include/clang/AST/DeclTemplate.h:2610]: Alright. Do you have any suggestions for the name?

File: clang/unittests/ASTMatchers/ASTMatchersTraversalTest.cpp (hunk 8 above)
Comment (erichkeane) [reply clang/unittests/ASTMatchers/ASTMatchersTraversalTest.cpp:6444]: Ping on this.

File: clang/include/clang/AST/DeclTemplate.h (hunk 10 above)
Comment (erichkeane) [reply clang/include/clang/AST/DeclTemplate.h:2610]: Not a good one :)  Perhpas: `TemplateArgListInfoOrExplicitSpecInfo` ?

File: clang/unittests/ASTMatchers/ASTMatchersTraversalTest.cpp (hunk 8 above)
Comment (sdkrystian) [reply clang/unittests/ASTMatchers/ASTMatchersTraversalTest.cpp]: @erichkeane I just pushed a commit which converts all the `#if`'d out test to use `hasTemplateArgument`/`hasAnyTemplateArgument` instead of `hasTemplateArgumentLoc`/`hasAnyTemplateArgumentLoc`. However, this is a temporary measure.

I want to add support for `ClassTemplateSpecializationDecl`, `VarTemplateSpecializationDecl`, and `FunctionDecl` to `hasTemplateArgumentLoc`/`hasAnyTemplateArgumentLoc` so these tests can be switched back to using the `TemplateArgumentLoc` matchers, but I'm not sure whether it should be included in this PR, or have it's own follow-up PR (I would prefer to include it in this one).

**Review by erichkeane — APPROVED** [review]

**Comment by sdkrystian** [comment]
Ping @AaronBallman

**Comment by erichkeane** [comment]
> Ping @AaronBallman

Aaron is on vacation this week, is there feedback from him that you're looking for?

**Comment by sdkrystian** [comment]
I'd like someone to review the AST matcher changes (the code owner listed in `CodeOwners.rst` appears to be inactive).

**Comment by erichkeane** [comment]
> I'd like someone to review the AST matcher changes (the code owner listed in `CodeOwners.rst` appears to be inactive).

Ah, yes!  Unfortunately I think he's the best person to review, I don't know of anyone familiar enough with the matchers.  But he'll be back next week.

**Comment by sdkrystian** [comment]
Ping @AaronBallman

File: clang/include/clang/ASTMatchers/ASTMatchers.h [hunk 13]
```diff
@@ -764,9 +764,9 @@ AST_POLYMORPHIC_MATCHER(isImplicit,
   return Node.isImplicit();
 }
 
-/// Matches classTemplateSpecializations, templateSpecializationType and
-/// functionDecl that have at least one TemplateArgument matching the given
-/// InnerMatcher.
+/// Matches templateSpecializationTypes, class template specializations,
```
Comment (AaronBallman) [inline clang/include/clang/ASTMatchers/ASTMatchers.h:767]: The AST matcher changes look correct to me, but you should re-run `clang/docs/tools/dump_ast_matchers.py` to regenerate the HTML documentation.

File: clang/include/clang/ASTMatchers/ASTMatchers.h [hunk 14]
```diff
@@ -4075,9 +4076,8 @@ AST_POLYMORPHIC_MATCHER_P(
     AST_POLYMORPHIC_SUPPORTED_TYPES(
         BlockDecl, CXXBaseSpecifier, CXXCtorInitializer, CXXFunctionalCastExpr,
         CXXNewExpr, CXXTemporaryObjectExpr, CXXUnresolvedConstructExpr,
-        ClassTemplateSpecializationDecl, CompoundLiteralExpr, DeclaratorDecl,
```
Comment (AaronBallman) [inline clang/include/clang/ASTMatchers/ASTMatchers.h:4078]: We should probably call this out as a potentially breaking change in the release notes because code that used to work with this matcher will need to change.

File: clang/include/clang/ASTMatchers/ASTMatchers.h (hunk 13 above)
Comment (sdkrystian) [reply clang/include/clang/ASTMatchers/ASTMatchers.h:767]: @AaronBallman Already did so :)

**Comment by sdkrystian** [comment]
@AaronBallman release note added (feel free to suggest any changes to it :))

File: clang/docs/ReleaseNotes.rst [hunk 15]
```diff
@@ -111,6 +111,13 @@ Clang Frontend Potentially Breaking Changes
     $ clang --target=<your target triple> -print-target-triple
     <the normalized target triple>
 
+- The ``getTypeAsWritten`` member function has been removed from ``ClassTemplateSpecializationDecl`` and
+  ``VarTemplateSpecializationDecl``, and a new member function ``getTemplateArgsAsWritten`` returning a
+  ``const ASTTemplateArgumentListInfo *`` has been added to access to the explicitly written template
+  argument list of explicit specializations, partial specializations, and explicit instantiations of
+  class and variable templates. Consequently, the ``hasTypeLoc`` AST matcher will no longer match these
+  declarations and existing uses should switch to ``TemplateArgumentLoc`` matchers instead.
+
```
Comment (AaronBallman) [inline clang/docs/ReleaseNotes.rst:120]: I think we can remove the internal details and go with something shorter, like:
```
The `hasTypeLoc` AST matcher will no longer match a `classTemplateSpecializationDecl`; existing uses should switch to `templateArgumentLoc` or `hasAnyTemplateArgumentLoc` instead.
```
WDYT?

**Review by AaronBallman — APPROVED** [review]
LGTM aside from a possible nit with the release note (I don't need to re-review).

File: clang/docs/ReleaseNotes.rst (hunk 15 above)
Comment (sdkrystian) [reply clang/docs/ReleaseNotes.rst:120]: > I think we can remove the internal details and go with something shorter, like:
> 
> ```
> The `hasTypeLoc` AST matcher will no longer match a `classTemplateSpecializationDecl`; existing uses should switch to `templateArgumentLoc` or `hasAnyTemplateArgumentLoc` instead.
> ```
> 
> WDYT?

@AaronBallman As a strictly AST matcher release note, I like your proposed wording. However, I think it's important to include the removal of `getTypeAsWritten` (and the addition of its replacement, `getTemplateArgsAsWritten`) so tooling authors know how to fix any code which is broken by these changes. Would adding two release notes be preferable? For example:

> - ``getTypeAsWritten`` has been removed from ``ClassTemplateSpecializationDecl`` and ``VarTemplateSpecializationDecl`` in favor of a new function ``getTemplateArgsAsWritten``.
> - The ``hasTypeLoc`` AST matcher will no longer match a ``classTemplateSpecializationDecl``; existing uses should switch to ``templateArgumentLoc`` or ``hasAnyTemplateArgumentLoc`` instead.

Comment (AaronBallman) [reply clang/docs/ReleaseNotes.rst:120]: Our C++ APIs are not stable across releases, so it's very typical for folks using Clang internals to be broken between releases (this is why we strongly encourage folks to work at Tip of Tree rather than on a release-by-release basis). That's why we don't usually add release notes for internal API changes.

Comment (sdkrystian) [reply clang/docs/ReleaseNotes.rst:120]: Alright, I'll just include the AST matcher release note then :)

**Comment by sdkrystian** [comment]
This seems to cause a crash in lldb in `ClassTemplateSpecializationDecl::getSourceRange()`... investigating

**Comment by adrian-prantl** [comment]
This broke several tests in the LLDB testsuite, I'm going to have to revert this.
https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/3480/

**Comment by kimgr** [comment]
@sdkrystian It looks like this PR broke IWYU with respect to this assumption:
> Moreover, we don't ever need the type as written -- in almost all cases, we only want the template arguments (e.g. in tooling use-cases).

As I understand it, IWYU did use the type as written to do resugaring of template types, here: https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu_ast_util.cc#L935

Do you have any tips for how to implement that on top of the new model? Manually type-switch all TemplateArgument kinds and map to a Type, possibly via Expr?

**Comment by sdkrystian** [comment]
@kimgr The linked code seems to only be concerned with the `SourceLocation` of the `Decl` (which should be the same `SourceLocation` returned by `Decl::getLocation`), and the template arguments as written (which can be accessed via `getTemplateArgsAsWritten` per this patch). I would imagine the fix here would be to handle explicit instantiations of class templates the same what you handle explicit instantiations of function templates (which also do not store a `TypeLoc` but can have their template arguments as written accessed in a similar manner).

**Comment by kimgr** [comment]
@sdkrystian Thank you! I think there's some more subtleties to it, but I don't really have my head around it yet. I'll first try to understand what we're trying to achieve and then ping back of I need help with something concrete.
